File: api/Validators/salesOrder.js
'use strict';

/**
 * api/Validators/salesOrder.js
 *
 * Joi schemas for the sales-orders resource. Mirrors quotation.js — see that
 * file's header comment for the rationale. Server-managed columns (id,
 * company_id, order_no when blank, all money totals, cgst/sgst/igst,
 * round_off, order_status, converted_invoice_id, status, tally_*, created_by,
 * timestamps, deleted_at) are intentionally NOT accepted from the body; the
 * totals are computed by SalesOrderController.computeTotals inside the write
 * transaction. NEVER trust client-sent totals — `.unknown(false)` (Joi's
 * default) drops them silently if sent.
 *
 * Schemas:
 *   createSalesOrderSchema — POST/PUT /sales-orders (customer_id required)
 *   listSalesOrderSchema   — GET  /sales-orders       (pagination + filters)
 */

const Joi = require('joi');

// Reusable optional positive-integer FK.
const fkId = Joi.number().integer().positive();
// Optional FK that also accepts an EXPLICIT null/'' (API clients send
// "location_id": null for "none" — omit-only would 422 on that).
const optFkId = fkId.allow(null, '');

// Reusable optional short/long text — trimmed, blank/null allowed.
const optText = (max) => Joi.string().trim().max(max).allow('', null);

/**
 * A single sales-order line. The money columns (taxable / gst_amount /
 * amount) are NOT accepted here — the controller computes them
 * authoritatively from quantity / rate / discount_pct / gst_rate /
 * tax_inclusive.
 */
const itemSchema = Joi.object({
    product_id:    fkId.allow(null, ''),
    description:   optText(2000),
    hsn:           optText(20),
    quantity:      Joi.number().greater(0).precision(2).required().messages({
        'number.base':     'Item quantity is required.',
        'number.greater':  'Item quantity must be greater than 0.',
        'any.required':    'Item quantity is required.',
    }),
    unit:          optText(30),
    rate:          Joi.number().min(0).precision(2).required().messages({
        'number.base':  'Item rate is required.',
        'number.min':   'Item rate must be 0 or more.',
        'any.required': 'Item rate is required.',
    }),
    discount_pct:  Joi.number().min(0).max(100).precision(2).default(0),
    gst_rate:      Joi.number().min(0).precision(2).default(0),
    godown:        optText(120),
    tax_inclusive: Joi.boolean().default(false),
});

const itemsArray = Joi.array().items(itemSchema).min(1).required().messages({
    'array.min':    'At least one line item is required.',
    'array.base':   'At least one line item is required.',
    'any.required': 'At least one line item is required.',
});

/**
 * POST /api/v1/sales-orders (also used for PUT — the shape is the same).
 * `customer_id` and at least one item are required; `order_no` is optional —
 * the controller generates one when blank.
 */
const createSalesOrderSchema = Joi.object({
    customer_id:     fkId.required().messages({
        'any.required':    'Customer is required.',
        'number.base':     'Customer is required.',
        'number.positive': 'Customer is required.',
    }),
    location_id:     optFkId,
    sales_person_id: optFkId,

    order_no:        optText(60),
    order_date:      Joi.date().iso().allow(null, ''),
    due_on:          Joi.date().iso().allow(null, ''),
    ledger_name:     optText(120),
    notes:           optText(2000),

    items:           itemsArray,
});

/**
 * GET /api/v1/sales-orders (query string)
 * Pagination + the filters the list handler reads. Param names match
 * listQuotationSchema (page/per_page/search/date_from/date_to) — the shared
 * web `apiList()` helper only understands those. `order_status` is the
 * sales-order-only delivery-lifecycle filter (pending/partially_delivered/
 * delivered/cancelled); it is deliberately NOT named `status` — that name is
 * reserved app-wide for the Tally-sync lifecycle.
 */
const listSalesOrderSchema = Joi.object({
    page:         Joi.number().integer().min(1).default(1),
    per_page:     Joi.number().integer().min(1).max(100).default(10),
    search:       Joi.string().trim().max(191).allow('', null),
    date_from:    Joi.date().iso().allow('', null),
    date_to:      Joi.date().iso().allow('', null),
    order_status: Joi.string().valid('pending', 'partially_delivered', 'delivered', 'cancelled', 'all').allow(''),
    mine:         Joi.alternatives().try(Joi.valid('1', 1, true, false), Joi.boolean()).allow('', null),
});

module.exports = {
    createSalesOrderSchema,
    listSalesOrderSchema,
};

File: app/lib/data/models/sales_order.dart
import 'voucher_item.dart';

/// One sales order from `GET /api/v1/sales-orders` (and its lines from
/// `GET /api/v1/sales-orders/:id`). Mirrors the `sales_orders` /
/// `sales_order_items` tables; pg returns numerics as strings, so the
/// coercions are defensive.
///
/// Two independent statuses ride on an order — do not conflate them:
///   • [orderStatus] — the DELIVERY lifecycle: pending / partially_delivered /
///     delivered / cancelled.
///   • [status]      — the Tally-sync lifecycle every voucher carries.
class SalesOrder {
  const SalesOrder({
    required this.id,
    required this.orderNo,
    this.customerId,
    this.customer,
    this.locationId,
    this.location,
    this.salesPersonId,
    this.orderDate,
    this.dueOn,
    this.ledgerName,
    this.subtotal,
    this.discount,
    this.taxable,
    this.cgst,
    this.sgst,
    this.igst,
    this.taxAmount,
    this.roundOff,
    this.total,
    this.orderStatus,
    this.convertedInvoiceId,
    this.status,
    this.notes,
    this.createdAt,
    this.items = const [],
  });

  final int id;
  final String orderNo;

  final int? customerId;
  final String? customer;
  final int? locationId;
  final String? location;
  final int? salesPersonId;

  final String? orderDate;

  /// When the goods are due to be delivered.
  final String? dueOn;
  final String? ledgerName;

  final num? subtotal;
  final num? discount;
  final num? taxable;
  final num? cgst;
  final num? sgst;
  final num? igst;
  final num? taxAmount;
  final num? roundOff;
  final num? total;

  /// pending | partially_delivered | delivered | cancelled.
  final String? orderStatus;

  /// Set once the order has been converted into a sales invoice.
  final int? convertedInvoiceId;

  /// Tally-sync lifecycle.
  final String? status;

  final String? notes;
  final String? createdAt;

  /// Only populated by the detail endpoint.
  final List<VoucherItem> items;

  bool get isConverted => convertedInvoiceId != null;

  factory SalesOrder.fromJson(Map<String, dynamic> j) => SalesOrder(
        id: _toInt(j['id']) ?? 0,
        orderNo: (j['order_no'] ?? '').toString(),
        customerId: _toInt(j['customer_id']),
        customer: _sn(j['customer']),
        locationId: _toInt(j['location_id']),
        location: _sn(j['location']),
        salesPersonId: _toInt(j['sales_person_id']),
        orderDate: _sn(j['order_date']),
        dueOn: _sn(j['due_on']),
        ledgerName: _sn(j['ledger_name']),
        subtotal: _toNum(j['subtotal']),
        discount: _toNum(j['discount']),
        taxable: _toNum(j['taxable']),
        cgst: _toNum(j['cgst']),
        sgst: _toNum(j['sgst']),
        igst: _toNum(j['igst']),
        taxAmount: _toNum(j['tax_amount']),
        roundOff: _toNum(j['round_off']),
        total: _toNum(j['total']),
        orderStatus: _sn(j['order_status']),
        convertedInvoiceId: _toInt(j['converted_invoice_id']),
        status: _sn(j['status']),
        notes: _sn(j['notes']),
        createdAt: _sn(j['created_at']),
        items: VoucherItem.listFrom(j['items']),
      );
}

/// Human label for the DELIVERY lifecycle (`order_status`).
String orderStatusLabel(String? s) {
  switch (s) {
    case 'pending':
      return 'Pending';
    case 'partially_delivered':
      return 'Partial';
    case 'delivered':
      return 'Delivered';
    case 'cancelled':
      return 'Cancelled';
    default:
      return s == null || s.isEmpty ? 'Pending' : s;
  }
}

String? _sn(Object? v) {
  if (v == null) return null;
  final s = v.toString().trim();
  return s.isEmpty ? null : s;
}

int? _toInt(Object? v) {
  if (v == null) return null;
  if (v is num) return v.toInt();
  final s = v.toString().trim();
  return s.isEmpty ? null : int.tryParse(s);
}

num? _toNum(Object? v) {
  if (v == null) return null;
  if (v is num) return v;
  final s = v.toString().trim();
  return s.isEmpty ? null : num.tryParse(s);
}
